### DISCUSSION TITLE
Low memory used and same time cost

### DISCUSSION THREAD

**Discussion by guanguoliang** [PHPOffice/PhpSpreadsheet#2515] [Ideas]
As my work department has a demand to export a lot of data for account statement, directly export excel is using large memory.My workmate tried to use “Memory saving” to reduce memory used.It did help to avoid use large memory, but spent double time to finish it which spent more than what we want.
So, I tried to change the code of version 1.12.0 is used in our group, and I get a way to use very low  memory as well as the time cost near the original way.
I want to submit a PR, however it seems a very old version, and what I change slightly destructs encapsulation, some repeated code by adjusting the logic and the order of export.It means the way to export can't use a easy method "save", it need to use more methods like "saveSheetHeader" , 'saveSheetFormData" etc. Because we need to control memory used and make sure the right format to create an Excel.
By the way ,because what we need is a normal excel, means I just support mergeCell, formula and style.But it can cover many large data export scene. Charts, media  and drawing nearly not exist.Ok, I do have no time to satisfy it.  (T_T)
I want to know if author @MarkBaker  has an interest to have this feature and I will try to add the feature in the newest version or maybe the old 1.12.0?

**Comment by MarkBaker** [comment]
If you have a way of saving memory, then I would indeed be interested; but it would need to support all the features that PHPSpreadsheet supports, and it would need to be PRed against the master branch. And any change to calling `save()` would be a BC break

**Reply by MarkBaker** [reply]
But can't you new "saveSheetHeader" , 'saveSheetFormData" etc. methods simply be called from the `save()` method? And I assume you're talking about one specific Writer, or could this change be applied to all Writers?

**Reply by guanguoliang** [reply]
Because I don't know about how the design of  Excel to be created, I just read the source code from PHPSpreadsheet, and try to be clever to separate `save()` into several methods to release memory, during a lot of data when we use `for()` or `foreach()` to load into each cell.Actually, I write the xml of zip in advance instead of loading all the data and finally using `save()` to write the whole file, that is how I save memory.So, I can't make sure to support all the features cause I use a special way to create and I don't know the details of creating an Excel.
Maybe this feature can just be used in the scene of big form to export, and if  we need all the features can use the original way.This is what I think about it.
I think I can use a specific Writer to achieve this feature, but I think it's not you want, emmmmm, I do the change of the master version first, and then you can see how it work, or maybe you can change the whole feature and with this low memory way.

**Reply by guanguoliang** [reply]
https://github.com/PHPOffice/PhpSpreadsheet/pull/2524
I probably can support all the features and still export with low memory, please CR this

**Reply by guanguoliang** [reply]
@MarkBaker if you have time, please help to review code, I make low memory to export come true, and I want to make this feature into  PHPSpreadsheet
